consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/FutureMessageBuffer.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import tech.pegasys.pantheon.ethereum.p2p.api.Message;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Buffer which holds future IBFT messages.
+ *
+ * <p>This buffer only allows messages to be added which have a chain height greater than current
+ * height and up to chain futureMessagesMaxDistance from the current chain height.
+ *
+ * <p>When the total number of messages is greater futureMessagesLimit then messages are evicted.
+ *
+ * <p>If there is more than one height in the buffer then all messages for the highest chain height
+ * are removed. Otherwise if there is only one height the oldest inserted message is removed.
+ */
+public class FutureMessageBuffer {
+  private final NavigableMap<Long, List<Message>> buffer;
+  private final long futureMessagesMaxDistance;
+  private final long futureMessagesLimit;
+
+  public FutureMessageBuffer(final long futureMessagesMaxDistance, final long futureMessagesLimit) {
+    this(futureMessagesMaxDistance, futureMessagesLimit, new TreeMap<>());
+  }
+
+  @VisibleForTesting
+  FutureMessageBuffer(
```

Comment:
Is this constructor really of use? ...looks like when it's called in the tests a `new TreeMap<>` is passed in

Reply:
It is, I'm using that so I can check the state of the map is correct in the tests without exposing the map to everyone. A new TreeMap<> is passed in, but I'm checking in my tests messages are added to the map correctly.

Reply:
That's a tight coupling between the test and the specific implementation of the logic, I'm not convinced that's the best approach.

Reply:
done. updated so don't need special constructor